dbms/src/Storages/MergeTree/ReplicatedMergeTreeCleanupThread.cpp

```diff
@@ -75,26 +75,62 @@ void ReplicatedMergeTreeCleanupThread::clearOldLogs()
     int children_count = stat.numChildren;
 
     /// We will wait for 1.1 times more records to accumulate than necessary.
-    if (static_cast<double>(children_count) < storage.data.settings.replicated_logs_to_keep * 1.1)
+    if (static_cast<double>(children_count) < storage.data.settings.min_replicated_logs_to_keep * 1.1)
         return;
 
     Strings replicas = zookeeper->getChildren(storage.zookeeper_path + "/replicas", &stat);
-    UInt64 min_pointer = std::numeric_limits<UInt64>::max();
+    UInt64 min_pointer_active_replica = std::numeric_limits<UInt64>::max();
+
+    Strings entries = zookeeper->getChildren(storage.zookeeper_path + "/log");
+
+    if (entries.empty())
+        return;
+
+    std::sort(entries.begin(), entries.end());
+
+    String min_saved_record_log_str = entries[std::max(0, entries.size() - storage.data.settings.max_replicated_logs_to_keep.value)];
```

Comment:
`entries.size()` and  `storage.data.settings.max_replicated_logs_to_keep.value` are both unsigned, so subtracting them is very dangerous.